### PR TITLE
Highsierra refactor

### LINKED ISSUE #53: Auth Restart during enablement doesn't work on High Sierra
I have performed no troubleshooting here, I am just tracking it here in case anyone wants to look into it.

### PR DESCRIPTION
**Pull request by weswhet**
#### New
* Adds the ability to Enable FileVault on HighSierra APFS without doing a restart or GUI prompt. Still, performs a restart on HFS Drives.
* Did a minor refactor of removing duplicated code and moving any shared functions to the Mechanisms file. 
* Logs now display the StandardError of `/usr/bin/fdesetup` when a failure occurs during enablement or key rotation. 
* Fixes #53 

#### Tests
Tested 10.12.4-6 virtual and physical
Tested 10.13 APFS and 10.13 HFS.

### CODE REVIEW

**Comment by weswhet** [comment]
@grahamgilbert any chance you could take a look at this.

**Comment by grahamgilbert** [comment]
This looks good. Let me test on a couple of machines.

**Comment by grahamgilbert** [comment]
* Looks fine on 10.12
* When installing in a 10.13 APFS VM the VM rebooted __without showing any UI__. I assume this isn't the intended behaviour.

**Comment by grahamgilbert** [comment]
Bump @weswhet

**Comment by weswhet** [comment]
Yeah sorry ran into this today. Working on a fix for Monday.

**Comment by weswhet** [comment]
@grahamgilbert should be good now. I assumed the `-authrestart` flag was ignored on APFS so I was including it regardless. That's what was causing the unexpected reboot. That addition is skipped now if on APFS.

**Comment by erikng** [comment]
@weswhet https://gist.github.com/erikng/18d97d188e58b94b4c31af0885e3f9b4 this should fix the issues with Xcode 9 and allow it to build.

**Comment by clburlison** [comment]
Part of this conversation happened in our #crypt channel in Slack. 

TLDR: @tburgin is the bomb and the following branch should be merged into this PR. https://github.com/tburgin/crypt/commits/mech-data

**Comment by weswhet** [comment]
I merged bur's additions to this to the dev branch
